<a href="https://colab.research.google.com/github/Ranjan4Kumar/Image_Classification_TF/blob/main/18th_Jan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import cv2
import os
import numpy as np
import random

In [17]:
mydir = '/content/drive/MyDrive/DS/TensorFlow_img_Classification_Project/train'

In [12]:
categories = ['cats','dogs']

In [18]:
mydir

'/content/drive/MyDrive/DS/TensorFlow_img_Classification_Project/train'

In [48]:
data = []

for i in categories:
  folder_path = os.path.join(mydir,i)

  if i=='cats':
    label = 0
  else:
    label = 1
  for j in os.listdir(folder_path):
    img_path = os.path.join(folder_path, j)
    img = cv2.imread(img_path)
    img = cv2.resize(img,(150,150))
    data.append([img,label])


In [40]:
random.shuffle(data)

In [49]:
X = []
Y = []

for i in data:
  X.append(i[0])
  Y.append(i[1])

In [50]:
Y = np.array(Y)
X = np.array(X)

In [51]:
X.shape

(3551, 150, 150, 3)

In [52]:
Y.shape

(3551,)

In [27]:
X = X/255

In [53]:
# from keras.models import Sequential
# from keras.layers import Conv2D, MaxPooling2D
# from keras.layers import Activation, Dropout, Flatten, Dense
import tensorflow as tf
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense, Conv2D, MaxPooling2D, Flatten, Activation,Dropout

# Creating CNN model now
# we will create 3 layers 1st will contain 32 filters 2nd will contain 64 filters and 3rd will contain 128 filters

model = Sequential()
# Our 1st Convolution layers
model.add(Conv2D(32, kernel_size=(3,3), input_shape=(150,150,3)))

model.add(Activation('relu'))
# we will add pooling layer
model.add(MaxPooling2D(pool_size = (2,2)))


model.add(Conv2D(32, kernel_size=(3,3)))
# we will add pooling layer
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size = (2,2)))



model.add(Conv2D(64, kernel_size=(3,3)))
# we will add pooling layer
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size = (2,2)))

In [54]:
model.add(Flatten())  # this converts our 3D feature maps to 1D feature vectors
model.add(Dense(64))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(1))
model.add(Activation('sigmoid'))

model.compile(loss='binary_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

In [55]:
model.fit(X,Y, epochs = 5, validation_split = 0.1)

Epoch 1/5
100/100 [==============================] - 10s 23ms/step - loss: 8.7025 - accuracy: 0.5524 - val_loss: 0.8448 - val_accuracy: 0.0000e+00
Epoch 2/5
100/100 [==============================] - 2s 18ms/step - loss: 0.7846 - accuracy: 0.5615 - val_loss: 0.7253 - val_accuracy: 0.0000e+00
Epoch 3/5
100/100 [==============================] - 2s 18ms/step - loss: 0.6888 - accuracy: 0.5621 - val_loss: 0.7468 - val_accuracy: 0.0000e+00
Epoch 4/5
100/100 [==============================] - 2s 18ms/step - loss: 0.6875 - accuracy: 0.5621 - val_loss: 0.7636 - val_accuracy: 0.0000e+00
Epoch 5/5
100/100 [==============================] - 2s 18ms/step - loss: 0.6867 - accuracy: 0.5621 - val_loss: 0.7789 - val_accuracy: 0.0000e+00


We can see over here Accuracy is about 56.21% 

### Now we will use Data Augumentation and we use same model 

In [56]:
from keras.preprocessing.image import ImageDataGenerator

batch_size = 16

# this is the augmentation configuration we will use for training
train_datagen = ImageDataGenerator(
        rescale=1./255,
        shear_range=0.2,
        zoom_range=0.2,
        horizontal_flip=True)

# this is the augmentation configuration we will use for testing:
# only rescaling
test_datagen = ImageDataGenerator(rescale=1./255)

# this is a generator that will read pictures found in
# subfolers of 'data/train', and indefinitely generate
# batches of augmented image data
train_generator = train_datagen.flow_from_directory(
        '/content/drive/MyDrive/DS/TensorFlow_img_Classification_Project/train',  # this is the target directory
        target_size=(150, 150),  # all images will be resized to 150x150
        batch_size=batch_size,
        class_mode='binary')  # since we use binary_crossentropy loss, we need binary labels

# this is a similar generator, for validation data
validation_generator = test_datagen.flow_from_directory(
        '/content/drive/MyDrive/DS/TensorFlow_img_Classification_Project/test',
        target_size=(150, 150),
        batch_size=batch_size,
        class_mode='binary')

Found 3551 images belonging to 2 classes.
Found 5009 images belonging to 2 classes.


In [57]:
# from keras.models import Sequential
# from keras.layers import Conv2D, MaxPooling2D
# from keras.layers import Activation, Dropout, Flatten, Dense
import tensorflow as tf
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense, Conv2D, MaxPooling2D, Flatten, Activation,Dropout

# Creating CNN model now
# we will create 3 layers 1st will contain 32 filters 2nd will contain 64 filters and 3rd will contain 128 filters

model = Sequential()
# Our 1st Convolution layers
model.add(Conv2D(32, kernel_size=(3,3), input_shape=(150,150,3)))

model.add(Activation('relu'))
# we will add pooling layer
model.add(MaxPooling2D(pool_size = (2,2)))


model.add(Conv2D(32, kernel_size=(3,3)))
# we will add pooling layer
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size = (2,2)))



model.add(Conv2D(64, kernel_size=(3,3)))
# we will add pooling layer
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size = (2,2)))

In [58]:
model.add(Flatten())  # this converts our 3D feature maps to 1D feature vectors
model.add(Dense(64))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(1))
model.add(Activation('sigmoid'))

model.compile(loss='binary_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])

In [59]:
# Whenever we are using datagenerator we will use fit_generator

model.fit_generator(
    train_generator,
    steps_per_epoch=3551 // batch_size,
        epochs=5,
        validation_data=validation_generator,
        validation_steps=800 // batch_size)


model.save_weights('first_try.h5')


Epoch 1/5


<ipython-input-59-548deb3dcea4>:3: UserWarning: `Model.fit_generator` is deprecated and will be removed in a future version. Please use `Model.fit`, which supports generators.
  model.fit_generator(


221/221 [==============================] - ETA: 0s - loss: 0.7063 - accuracy: 0.5406

KeyboardInterrupt: ignored